frame/pools/src/lib.rs

```diff
@@ -0,0 +1,1135 @@
+	pub enum Event<T: Config> {
+		Joined { delegator: T::AccountId, pool: T::AccountId, bonded: BalanceOf<T> },
+		PaidOut { delegator: T::AccountId, pool: T::AccountId, payout: BalanceOf<T> },
+		Unbonded { delegator: T::AccountId, pool: T::AccountId, amount: BalanceOf<T> },
+		Withdrawn { delegator: T::AccountId, pool: T::AccountId, amount: BalanceOf<T> },
+	}
+
+	#[pallet::error]
+	#[cfg_attr(test, derive(PartialEq))]
+	pub enum Error<T> {
+		/// A (bonded) pool id does not exist.
+		PoolNotFound,
+		/// An account is not a delegator.
+		DelegatorNotFound,
+		/// A reward pool does not exist. In all cases this is a system logic error.
+		RewardPoolNotFound,
+		/// A sub pool does not exist.
+		SubPoolsNotFound,
+		/// An account is already delegating in another pool. An account may only belong to one
+		/// pool at a time.
+		AccountBelongsToOtherPool,
+		/// The pool has insufficient balance to bond as a nominator.
+		InsufficientBond,
+		/// The delegator is already unbonding.
+		AlreadyUnbonding,
+		/// The delegator is not unbonding and thus cannot withdraw funds.
+		NotUnbonding,
+		/// Unbonded funds cannot be withdrawn yet because the bond duration has not passed.
+		NotUnbondedYet,
+		/// The amount does not meet the minimum bond to start nominating.
+		MinimumBondNotMet,
+		/// The transaction could not be executed due to overflow risk for the pool.
+		OverflowRisk,
+		/// An error from the staking pallet.
+		StakingError,
+		// Likely only an error ever encountered in poorly built tests.
+		/// A pool with the generated account id already exists.
+		IdInUse,
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Join a pre-existing pool.
+		///
+		/// Notes
+		/// * an account can only be a member of a single pool.
+		/// * this will *not* dust the delegator account, so the delegator must have at least
+		///   `existential deposit + amount` in their account.
+		#[pallet::weight(666)]
+		pub fn join(
+			origin: OriginFor<T>,
+			amount: BalanceOf<T>,
+			target: T::AccountId,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			// If a delegator already exists that means they already belong to a pool
+			ensure!(!Delegators::<T>::contains_key(&who), Error::<T>::AccountBelongsToOtherPool);
+
+			let mut bonded_pool = BondedPools::<T>::get(&target).ok_or(Error::<T>::PoolNotFound)?;
+			bonded_pool.ok_to_join_with(&target, amount)?;
+
+			// The pool should always be created in such a way its in a state to bond extra, but if
+			// the active balance is slashed below the minimum bonded or the account cannot be
+			// found, we exit early.
+			if !T::StakingInterface::can_bond_extra(&target, amount) {
+				Err(Error::<T>::StakingError)?;
+			}
+
+			// We don't actually care about writing the reward pool, we just need its
+			// total earnings at this point in time.
+			let reward_pool = RewardPools::<T>::get(&target)
+				.ok_or(Error::<T>::RewardPoolNotFound)?
```

Comment:
this should be defensive: if the pool exists, it must also have a reward pool, right? 

Generally, you should try and write down all the invariants of a "Pool as a whole" (i.e. a pool and all of its sub-pools)

Reply:
202153f05ba53e5b4db79091252f55885489a2b8